lib/lint.ml

```diff
@@ -10,10 +28,99 @@ module Checks = struct
     | DuneConstraintMissing
     | DuneIsBuild
     | BadDuneConstraint of string * string
+    | UnexpectedFile of string
+    | ForbiddenPerm of string
+    | OpamLint of (int * [ `Warning | `Error ] * string)
     | FailedToDownload of string
+    | NameCollision of string
     | WeakChecksum of string
     | PinDepends
     | ExtraFiles
+    | RestrictedPrefix of string
+    | PrefixConflictClassMismatch of prefix_conflict_class_mismatch
+
+  module Prefix = struct
+    (* For context, see https://github.com/ocurrent/opam-repo-ci/pull/316#issuecomment-2160069803 *)
+    let prefix_conflict_class_map =
+      [
+        ("mysys2-", "msys2-env");
+        ("arch-", "ocaml-arch");
+        ("ocaml-env-mingw", "ocaml-env-mingw");
+        ("ocaml-env-msvc", "ocaml-env-msvc");
+        ("host-arch-", "ocaml-host-arch");
+        ("host-system-", "ocaml-host-system");
+        ("system-", "ocaml-system");
+      ]
+
+    let conflict_class_prefix_map =
+      List.map (fun (a, b) -> (b, a)) prefix_conflict_class_map
+
+    let prefixes = List.map fst prefix_conflict_class_map
+
+    let check_name_restricted_prefix ~pkg _opam =
+      let name = OpamPackage.name_to_string pkg in
+      List.filter_map
+        (fun prefix ->
+          if String.starts_with ~prefix name then
+            Some (pkg, RestrictedPrefix prefix)
+          else None)
+        prefixes
+
+    let check_prefix_without_conflict_class ~pkg name conflict_classes =
+      let prefix =
+        List.find_opt (fun prefix -> String.starts_with ~prefix name) prefixes
+      in
+      match prefix with
+      | None -> []
+      | Some prefix -> (
+          match List.assoc_opt prefix prefix_conflict_class_map with
+          | None ->
+              print_endline
+              @@ Printf.sprintf
+                   "BUG: prefix '%s' not found in conflict class map" prefix;
+              []
```

Comment:
iiuc, in the context of our CLI tool, this should probably raise an exception, as it would mean a logic error in the program. WDYT?

Reply:
yes, that makes sense. I've updated the code to error out.